pkg/eval/math/math.go

```diff
@@ -0,0 +1,288 @@
+// ```elvish-transcript
+// ~> math:is-inf 123
+//
+// ~> math:is-inf 123 0
+// ▶ $false
+// ~> math:is-inf (float64 inf) 1
+// ▶ $true
+// ~> math:is-inf (float64 inf) -1
+// ▶ $false
+// ~> math:is-inf (float64 -inf) -1
+// ▶ $true
+// ```
+
+//elvdoc:fn is-nan
+//
+// ```elvish
+// math:is-nan $number $sign
+// ```
+//
+// If the number is not-a-number (NaN) output `$true` else `$false`.
+//
+// ```elvish-transcript
+// ~> math:is-nan 123
+// ▶ $false
+// ~> math:is-nan (float64 inf)
+// ▶ $false
+// ~> math:is-nan (float64 nan)
+// ▶ $true
+// ```
+
+//elvdoc:fn log
+//
+// ```elvish
+// math:log $number
+// ```
+//
+// Compute the natural logarithm of a number. Example:
+//
+// ```elvish-transcript
+// ~> math:log 1.0
+// ▶ (float64 1)
+// ~> math:log -2.3
+// ▶ (float64 NaN)
+// ```
+
+//elvdoc:fn log10
+//
+// ```elvish
+// math:log10 $number
+// ```
+//
+// Compute the base 10 logarithm of a number. Example:
+//
+// ```elvish-transcript
+// ~> math:log10 100.0
+// ▶ (float64 2)
+// ~> math:log10 -1.7
+// ▶ (float64 NaN)
+// ```
+
+//elvdoc:fn log2
+//
+// ```elvish
+// math:log2 $number
+// ```
+//
+// Compute the base 2 logarithm of a number. Example:
+//
+// ```elvish-transcript
+// ~> math:log2 8
+// ▶ (float64 3)
+// ~> math:log2 -5.3
+// ▶ (float64 NaN)
+// ```
+
+//elvdoc:fn round
+//
+// ```elvish
+// math:round $number
+// ```
+//
+// Output the nearest integer, rounding half away from zero.
+//
+// ```elvish-transcript
+// ~> math:round -1.1
+// ▶ (float64 -1)
+// ~> math:round 2.5
+// ▶ (float64 3)
+// ```
+
+//elvdoc:fn round-to-even
+//
+// ```elvish
+// math:round-to-even $number
+// ```
+//
+// Output the nearest integer, rounding ties to even.
+//
+// ```elvish-transcript
+// ~> math:round-to-even -1.1
+// ▶ (float64 -1)
+// ~> math:round-to-even 2.5
+// ▶ (float64 2)
+// ```
+
+//elvdoc:fn sin
+//
+// ```elvish
+// math:sin $number
+// ```
+//
+// Compute the sine of a number in radians (not degress). Example:
+//
+// ```elvish-transcript
+// ~> math:sin 0
+// ▶ (float64 0)
+// ~> math:sin 3.14159265
+// ▶ (float64 3.5897930298416118e-09)
+// ```
+
+//elvdoc:fn sqrt
+//
+// ```elvish
+// math:sqrt $number
+// ```
+//
+// Compute the square-root of a number. Example:
+//
+// ```elvish-transcript
+// ~> math:sqrt 0
+// ▶ (float64 0)
+// ~> math:sqrt 4
+// ▶ (float64 2)
+// ~> math:sqrt -4
+// ▶ (float64 NaN)
+// ```
+
+//elvdoc:fn tan
+//
+// ```elvish
+// math:tan $number
+// ```
+//
+// Compute the tangent of a number in radians (not degress). Example:
+//
+// ```elvish-transcript
+// ~> math:tan 0
+// ▶ (float64 0)
+// ~> math:tan 3.14159265
+// ▶ (float64 -0.0000000035897930298416118)
+// ```
+
+//elvdoc:fn trunc
+//
+// ```elvish
+// math:trunc $number
+// ```
+//
+// Output the integer value of the float64.
+//
+// ```elvish-transcript
+// ~> math:trunc -1.1
+// ▶ (float64 -1)
+// ~> math:trunc 2.5
+// ▶ (float64 2)
+// ```
+
+// Ns is the namespace for the math: module.
+var Ns = eval.NewNs().AddGoFns("math:", fns)
```

Comment:
I like, and use, method chaining in a lot of Python code I deal with. I am not a fan of lengthy method chaining. It tends to obscure what is happening. Which makes it harder to understand the code. I appreciate that at the moment the alternative formulation you propose is easy to understand. The problem is what happens when additional constants, methods, etc., are chained together. Which will almost certainly occur. You've objected to the use of `init()` functions before but I don't understand why since, AFAICT, they don't introduce any performance or correctness problems when used in this manner. The mixing of `eval.Ns{...}` declarations with `eval.NewNs()`

Nonetheless, to facilitate getting this merged I've made the suggested change.

P.S., There are just four uses of `NewNs()`. Compared to twenty-four uses of `Ns{}`. Should the former be converted to the latter form?

Reply:
**Re use of method chaining**:

The problem with `init` is that it does not detect cyclic dependencies in initialization. Compare the following two pieces of code:

Using `init`:

```go
// In file a.go
var a int
func init() { a = b + 1 }

// In file b.go
var b int
func init() { b = a + 1}
```

Not using `init`:

```go
// In file a.go
var a int = b + 1

// In file b.go
var b int = a + 1 
```

Go will correctly point out that the second version has an initialization loop, but it will happily compile the first version. Now, what should the values of `a` and `b` be? :)

It is often a matter of trade-off, of course. Sometimes using `init` is much cleaner than the alternative that avoids it. But in this case, my call is that method chaining is the lesser evil here.

**Re use of NewNs**: Yes, they can be all replaced by `Ns{}`.

Reply:
I understand your point about an indeterminate initialization sequence. That is inherent in something like Go's module level `init()` mechanism. I was not aware that the compiler would error out if that happens in an example such as you showed. I'll explore that feature. Nonetheless, I still maintain that using the `init()` mechanism is clearer in this situation for non-trivial initialization and, when used in a stylized fashion with an understanding of the possible indeterminacy, is easier to understand.